codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/MiddlewareRegistrar.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen.integration;
+
+import java.util.Objects;
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.go.codegen.SymbolUtils;
+import software.amazon.smithy.utils.SmithyBuilder;
+import software.amazon.smithy.utils.ToSmithyBuilder;
+
+
+public class MiddlewareRegistrar implements ToSmithyBuilder<MiddlewareRegistrar> {
+    private final Symbol resolvedFunction;
+    private final Symbol functionArgument;
+
+    public MiddlewareRegistrar(Builder builder) {
+        this.resolvedFunction = builder.resolvedFunction;
+        this.functionArgument = builder.functionArgument;
+    }
+
+    /**
+     * @return Returns symbol that resolves to a function.
+     */
+    public Symbol getResolvedFunction() {
+        return resolvedFunction;
+    }
+
+    /**
+     * @return Returns a symbol denoting the argument of the resolved function.
+     */
+    public Symbol getFunctionArgument() {
+        return functionArgument;
+    }
```

Comment:
This would be more flexible as a list of arguments instead of a single argument. This would also capture the idea of zero arguments too.

Reply:
Updated to take a list of arguments